features/steps/givens/attributes.py

```diff
@@ -91,27 +87,18 @@ def step_impl(context, inst, attribute, tail="single"):
 def step_impl(context, inst):
     return [segments[-1] for curve in inst for segments in curve]
 
-@validate_step('Its final {segment_type}')
-def step_impl(context, segment_type):
-    context._push()
-    context.instances = list()
+
+@gherkin_ifc.step('Its final {segment_type}')
+def step_impl(context, inst, segment_type):
     business_logic_types = [f"IFCALIGNMENT{_}SEGMENT" for _ in ["HORIZONTAL", "VERTICAL", "CANT"]]
-    if segment_type.upper() == "SEGMENT":
-        # processing an ALS rule
-        for curves in context._stack[1]["instances"]:
-            for curve in curves:
-                for segments in curve:
-                    last_seg = segments[-1]
-                    context.instances.append(last_seg)
-                    setattr(context, 'applicable', True)
-    elif segment_type.upper() in business_logic_types:
+    if segment_type == "segment":
+        return [segments[-1] for curve in inst for segments in curve]
```

Comment:
1.

I don't think we should `return` and `yield` in the same function. This has a high potential of confusing. In python 2.x this wasn't even allowed.

2.

It seems to me that 

- @validate_step('Its final {segment_type}')
- @validate_step('Its final segment')

are redundant.

Reply:
Agree. The former was removed in a different edit and is no longer present.

Reply:
As it turned out the best solution was two separate implementation functions: one for ALS015 and a separate one for ALB015.